src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt

```diff
@@ -104,15 +117,52 @@ class HMPPSAuthService(
         hmppsAuthApiClient.get(detailUrl)
           .retrieve()
           .bodyToMono(UserDetailResponse::class.java)
+          .withRetryPolicy()
           .map { it.name.substringBefore(' ') },
         hmppsAuthApiClient.get(emailUrl)
           .retrieve()
           .onStatus({ it.equals(HttpStatus.NO_CONTENT) }, { Mono.error(UnverifiedEmailException()) })
           .bodyToMono(UserEmailResponse::class.java)
+          .withRetryPolicy()
           .map { it.email }
       )
         .map { UserDetail(it.t1, it.t2) }
         .block()
     }
   }
+
+  fun <T> Flux<T>.withRetryPolicy(): Flux<T> {
+    return this
+      .retryWhen(
+        Retry.max(maxRetryAttempts)
+          .filter { isTimeoutException(it) }
+          .doBeforeRetry { logRetrySignal(it) }
+      )
+  }
+
+  fun <T> Mono<T>.withRetryPolicy(): Mono<T> {
+    return this
+      .retryWhen(
+        Retry.max(maxRetryAttempts)
+          .filter { isTimeoutException(it) }
+          .doBeforeRetry { logRetrySignal(it) }
+      )
+  }
+
+  fun isTimeoutException(it: Throwable): Boolean {
+    // Timeout for NO_RESPONSE is wrapped in a WebClientRequestException
+    return it is ReadTimeoutException || it is ConnectTimeoutException ||
```

Comment:
`io.netty.handler.timeout.TimeoutException` and `java.net.ConnectException` are probably change-safe "intent-specific" superclasses

Reply:
They are very netty specific and hang off io.netty.channel.ChannelException which in turn extends RuntimeException.